### ISSUE TITLE
HardPhongShader and SoftPhongShader works with no difference for training

### ISSUE THREAD

**Issue by xdfeng07** [facebookresearch/pytorch3d#643]
The pytorch3d version i used is v0.3.0

And the training results have no difference between HardPhongShader and SoftPhongShader, the gradients seem to be able to propagate no matter whether HardPhongShader or SoftPhongShader is used

But i remembered that the HardPhongShader is supposed to be not differentiable, why it still works?

**Comment by FavorMylikes** [comment]
@xdfeng07 
Hello, may I ask you why you are saying that, is there any paper or document to make an explanation?
> But I remembered that the HardPhongShader is supposed to be not differentiable, why it still works?

**Comment by kjchalup** [comment]
This is an old task but maybe I can help. I wish the original question had a full repro example! I'm sure that for some tasks, you can use HardPhongShader. For example, it should work if you have a fixed mesh and camera pose, and only optimize RGB values or some other vertex attributes.

But in many other cases (e.g. mesh optimization with RGB colors) I'd think the training wouldn't converge -- though it still might. The reason is quite subtle, because indeed the gradients flow through both shaders. But HardPhongShader will produce a ton of discontinuities.

HardPhongShader uses the differentiable rasterizer to get pix_to_face and bary_coords, then uses the phong shader to get colors for each relevant bary_coord on the mesh (this step is differentiable too); finally, it uses the hard_rgb_blend blender to "blend" the multiple faces that pix_to_face assigned to each pixel. Except, hard_rgb_blend just picks the top face and ignores the others.

SoftPhongShader is the same in the first two stages. But in the last stage, it uses a blending algorithm that outputs a color that is a weighted sum of colors of all the interpolated face attributes that project to the given pixel.

Now note the RasterizationSettings.blur_radius parameter. It makes the rasterizer pretend that each mesh face has edges that "stick out". Now imagine triangle1 that is a direct neighbor of triangle2, and consider a point within triangle1 but very close to their boundary. If blur_radius > 0, triangle2's edge will "stick out" over the boundary, and the rasterizer might assign both a point within triangle1 and triangle2 to the same pixel. Now, SoftPhongShader will assign to that pixel a weighted sum of those two triangles' colors (at their bary_coords that correspond to the pixel's location).

Taken together, this means that SoftPhongShader is differentiable at triangle boundaries -- if you do e.g. mesh optimization and shift the boundary just a tiny bit, then for HardPhongShader you will have a discontinuity where a pixel's color changed suddenly. But in SoftPhongShader you will not get a discontinuity, because at the boundaries pixel colors are a weighted sum of values on the triangles on both sides of the boundary. Finally, the same is true for occlusion boundaries (object self-occlusions, as well as object-bg occlusions).

See Shichen Liu et al, 'Soft Rasterizer: A Differentiable Renderer for Image-based 3D Reasoning' for more info on the blending algorithm, and some illustrations.

**Comment by yhu9** [comment]
According to Figure 2b on "Accelerating 3D Deep Learning with PyTorch3D", the issue described by @kjchalup is exactly what pytorch3D aims to resolve.

My question is why does PyTorch3D have the hard shader variant of the soft shaders? Is it maybe for visualization purposes?

**Comment by bottler** [comment]
There are many reasons you might want the hard variant. One example already given is that you might only be trying to optimize colors.
